### ISSUE TITLE
Support configuration of service endpoints

### ISSUE THREAD

**Issue by timoguin** [cloudquery/cloudquery#1381]
I would like to run CloudQuery against alternative endpoints, primarily for testing purposes. This would allow running CloudQuery against something like [localstack](https://github.com/localstack/localstack) to accelerate testing.

To facilitate this, the AWS SDKs allow supplying a map of service names to endpoints. To expose this option to CloudQuery, the AWS provider would need it's config schema updated to allow specifying this mapping, and the client initialization logic will need to support the [EndpointResolver](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/aws#EndpointResolver).

The Terraform AWS provider is a great reference for this. It allows configuration such as the following:

```
provider "aws" {
  endpoints {
    dynamodb = "http://localhost:4569"
    s3       = "http://localhost:4572"
  }
}
```

### References

- https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/endpoints/
- https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/custom-service-endpoints

**Comment by timoguin** [comment]
@yevgenypats @erezrokah @bbernays

Given lack of movement (no interest?) in this feature, I thought I would ask for some guidance on getting a populated database for testing. I know there is good test coverage that utilizes fixtures and mocks, as well as integration tests. How could I tap into that easily? I just want to bring up a Postgres database that has been populated by these tests, that way I can shorten the dev cycle by working locally instead of having to worry about real resources and live data.

**Comment by yevgenypats** [comment]
Hi @timoguin Sorry for dropping the ball on that I wasn't sure if there were any follow-up tasks to do here as I thought the Postgraphile and Postgrest solves it.

I wanted to verify that I understand your question/request: Are you looking for a way to populate the mock data to a Postgres database so you can do some other testing for external workload without the need to populate the postgres db manually ?

**Comment by timoguin** [comment]
> Hi @timoguin Sorry for dropping the ball on that I wasn't sure if there were any follow-up tasks to do here as I thought the Postgraphile and Postgrest solves it.
> 
> I wanted to verify that I understand your question/request: Are you looking for a way to populate the mock data to a Postgres database so you can do some other testing for external workload without the need to populate the postgres db manually ?

First off, I appreciate the blazing fast response, @yevgenypats! There are a few different needs at play here. I'll break them down, and then we can update the body of this issue and/or create others once what I'm trying to do is more clear.

# Goals

## Alternate Endpoint Support

The AWS SDKs support the concept of _alternate endpoints_. This allows you to point them to either non-AWS endpoints (like motoserver or some other AWS-compatible API) or AWS's FIPS-compliant endpoints.

CloudQuery needs to expose these configuration options in the source plugin. That's the main reason I opened this issue. The Terraform AWS provider has a good implementation that can be referenced.

## Shorten Dev Cycle

When testing any solution that utilizes CloudQuery data (e.g., an API), I need to be able to have quick and reliable test data for any of the providers I utilize. I don't want to have to run sync operations against live infrastructure to see what a populated database looks and how it behaves. I need to be able to _quickly_ get intimately familiar with the data.

My first approach was to see if I could point CloudQuery at a localstack instance. With localstack, I can provision a bunch of fake infrastructure _very_ quickly, in a repeatable way that will also work in CI, so my thought was that I could run sync against that to populate the database. That wasn't possible due to lack of support for alternate endpoint.

## What about other source plugins?

While we can add support for alternate endpoints to the AWS source plugin, that only solves this issue for AWS, since the power of localstack is fairly unmatched in other providers.

What _would_ improve the situation is being able to quickly generate test data for any available source plugin. I figured the best source for that data was the test suite.

# Action Items

- Add support for the alternate endpoints feature supported by the AWS SDK (this issue)
- Add support for FIPS-compliant AWS endpoints Create an issue to add FIPS support to the AWS source plugin
- Enable an easy, straightforward way to generate test data and/or populate a test database

FIPS support depends on the AWS source plugin exposing alternate endpoint configuration options. The test data.

# 💡 Other Ideas

We might want to consider adding a command or flag(s) to the CLI that would allow it to handle generation of test data. 

Potential options:

- `cloudquery generate`
- `cloudquery sync --test-data`
- `cloudquery sync --from-fixtures`

I'm not sure how feasible that would be for the CLI to do, but I need _something_ that can get me test data outside the context of CloudQuery's internal test suite.

Ideas are welcome.

**Comment by yevgenypats** [comment]
Thanks for this great write-up! Few thoughts on top of my head.

1) Re the AWS endpoint, I need to take a look and refresh my memory on what is currently possible in the AWS SDK but I do know there are some limitations such as this issue: https://github.com/aws/aws-sdk/pull/230 . I don't have an answer yet on this part but I'll go back do some research and discuss with the team.
2) Re generating test data I think in general that seems to be something out of scope for CloudQuery and I would assume there are tools that specialised at this. I did a quick Github/Google search and stumbled upon this: https://github.com/emirozer/fake2db. 
3) other ideas: Especially for other plugins where there is no localstack things that can generalise well is either tools that generate fake data according to schema and/or some kind of snapshot export/load so you sync once and then dump data and then load it again (I know there are also some tools that clean up various sensitive info from production data if you are concerned about it in testing scenarios so you have "clean" snapshots).

Would love to hear your thoughts and if something of the above will work for your use-case?

**Comment by timoguin** [comment]
> 1. Re the AWS endpoint, I need to take a look and refresh my memory on what is currently possible in the AWS SDK but I do know there are some limitations such as this issue: [Add proposal for configuring endpoints per service aws/aws-sdk#230](https://github.com/aws/aws-sdk/pull/230) . I don't have an answer yet on this part but I'll go back do some research and discuss with the team.

That issue is probably why I haven't contributed a PR to add this feature, but I hadn't seen it before. (Thanks for that. I've subscribed to it now.)

I initially thought this feature would only be a few lines of code to change. When I dug more into how the Terraform AWS provider handles this, I found more logic in the background related to mapping and resolving those different service endpoints, so I took a step back. I do know that the UX with Terraform does it is very nice. It's trivial to point the provider to many different possible endpoints, so I can locally test things like Terraform modules without actually hitting AWS.

> 2. Re generating test data I think in general that seems to be something out of scope for CloudQuery and I would assume there are tools that specialised at this. I did a quick Github/Google search and stumbled upon this: https://github.com/emirozer/fake2db.
> 3. other ideas: Especially for other plugins where there is no localstack things that can generalise well is either tools that generate fake data according to schema and/or some kind of snapshot export/load so you sync once and then dump data and then load it again (I know there are some companies that deal with those challenges: https://www.snaplet.dev/ - didn't have a chance to use it but met with the folks a while back).

I'll look into this more, but my initial thought is that it adds a lot of overhead. It seems like I'd have to script out something that fully understands all of the CloudQuery schemas and their relationships to actually generate anything useful. Then I would also have an undue burden of updating that logic any time upstream CloudQuery schemas change.

I want to get _closer_ to the upstream development cycle, not further away.

> Would love to hear your thoughts and if something of the above will work for your use-case?

My thought process in asking about the test data was that we should be able to tap into CloudQuery's actual unit and/or integration test suite for this. So if the integration tests are standing up and populating a local Postgres database, for example, we could maybe utilize that in some way instead of just tearing it down at the end of the test runs.

If _any_ of this kind of data is generated and/or available to the test suite, I think it would be best if I could hook into that somehow. Since the repositories have been reorganized with the 1.0 release, I'm having to dig around a bit more to find all the pieces I was looking at before, so any guidance would be appreciated.

In building something that relies completely on CloudQuery and the data it generates, I want to be able to quickly and easily run full end-to-end testing with all components. CloudQuery's test suite tests CloudQuery. Mine would test the full infrastructure involved with the solution. I want to know _quickly_ if something from upstream affects our services. Win-win for all of us. :)

**Comment by timoguin** [comment]
Another thing about motoserver and localstack is that they expose a single endpoint for all services. They handle the routing internally, so I don't think there should be any requirements for multiple service-specific endpoints (although that capability is ideal).

**Comment by yevgenypats** [comment]
> Another thing about motoserver and localstack is that they expose a single endpoint for all services. They handle the routing internally, so I don't think there should be any requirements for multiple service-specific endpoints (although that capability is ideal).

That's good to know. I'll take a look at this.

Few notes on our tests: Our mock tests generates random data (with the correct type) because how our code is built there is no place where a bug can be introduced due to data not being a real world data as long as it's the same type. The TLDR reason for that is the following flow:
1) AWS SDK: This validates whatever is returned from the server and then expose the data in strongly typed go
2) our mock tests make sure that every field that is populated in those structs will populate the cloudquery schema (the data inside each field doen't matter as long as it's get populated)
3) Given CloudQuery schema is strongly typed as well this ensures that whatever was converted to CQ schema will be inserted successfully to any of the supported databases.

This is why my first suggestion was to test your application with random data that conforms to the PostgreSQL schema (as we do basically the same but you don't need to hook into our tests as it's most prob easier to create random data directly to PostgreSQL as you want to have e2e tests). 

In any case I do understand the private endpoint feature and use-cases so I'll try to take a look at how this implemented in terraform.
Also, if that of any help I can jump on a call and we can discuss/brainstorm some other potential strategies or best practices.

**Comment by timoguin** [comment]
> In any case I do understand the private endpoint feature and use-cases so I'll try to take a look at how this implemented in terraform.

Much appreciated! 🙏

> Also, if that of any help I can jump on a call and we can discuss/brainstorm some other potential strategies or best practices.

I would be happy to do a call. We could code dive the Terraform provider together if you'd like. I looked through everything back when I first opened this issue, so I'm familiar with the code base.

**Comment by bbernays** [comment]
@timoguin - We have added the ability to override the `endpoint_url` via the spec file. While the changes have been merged to the main branch the functionality will be released sometime early next week in `v9.0.0` pr [can be found here](https://github.com/cloudquery/cloudquery/pull/5872).

During the development of the feature we tested it against LocalStack and it mostly worked found some issues with the data that LocalStack returns... We are working on a tutorial with how to do this setup so that the rest of the community can benefit. You can check out that [draft here](https://github.com/cloudquery/cloudquery/pull/5946)

We would love to hear more about how your work with CQ and LocalStack goes either on a GH issue or on [our discord](https://www.cloudquery.io/discord)
